prov/lnx/src/lnx_domain.c

```diff
@@ -0,0 +1,584 @@
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include "config.h"
+
+#include <assert.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <dirent.h>
+#include <ctype.h>
+
+#include <rdma/fi_errno.h>
+#include "ofi_util.h"
+#include "ofi.h"
+#include "ofi_str.h"
+#include "ofi_prov.h"
+#include "ofi_perf.h"
+#include "ofi_hmem.h"
+#include "rdma/fi_ext.h"
+#include "lnx.h"
+
+static struct fi_ops_domain lnx_domain_ops = {
+	.size = sizeof(struct fi_ops_domain),
+	.av_open = lnx_av_open,
+	.cq_open = lnx_cq_open,
+	.endpoint = lnx_endpoint,
+	.scalable_ep = lnx_scalable_ep,
+	.cntr_open = fi_no_cntr_open,
+	.poll_open = fi_no_poll_open,
+	.stx_ctx = fi_no_stx_context,
+	.srx_ctx = fi_no_srx_context,
+	.query_atomic = fi_no_query_atomic,
+	.query_collective = fi_no_query_collective,
+};
+
+static int lnx_cleanup_domains(struct local_prov *prov)
+{
+	int rc, frc = 0;
+	struct local_prov_ep *ep;
+
+	dlist_foreach_container(&prov->lpv_prov_eps,
+				struct local_prov_ep, ep, entry) {
+		if (!ep->lpe_domain)
+			continue;
+		rc = fi_close(&ep->lpe_domain->fid);
+		if (rc)
+			frc = rc;
+	}
+
+	return frc;
+}
+
+static int lnx_domain_close(fid_t fid)
+{
+	int rc = 0;
+	struct local_prov *entry;
+	struct lnx_domain *domain;
+
+	domain = container_of(fid, struct lnx_domain, ld_domain.domain_fid.fid);
+
+	/* close all the open core domains */
+	dlist_foreach_container(&domain->ld_fabric->local_prov_table,
+				struct local_prov,
+				entry, lpv_entry) {
+		rc = lnx_cleanup_domains(entry);
+		if (rc)
+			FI_INFO(&lnx_prov, FI_LOG_CORE, "Failed to close domain for %s\n",
+					entry->lpv_prov_name);
+	}
+
+	ofi_mr_cache_cleanup(&domain->ld_mr_cache);
+
+	rc = ofi_domain_close(&domain->ld_domain);
+
+	free(domain);
+
+	return rc;
+}
+
+static int
+lnx_mr_regattrs_all(struct local_prov *prov, const struct fi_mr_attr *attr,
+		    uint64_t flags, struct lnx_mem_desc_prov *desc)
+{
+	int rc = 0;
+	struct local_prov_ep *ep;
+
+	desc->prov = prov;
+
+	/* TODO: This is another issue here because MR registration can happen
+	 * quiet often
```

Comment:
My worry is that if we're calling mr_regattr before every send/recv, this could be a performance hit. I'm flagging it here so I don't forget to analyze the performance with/without it to understand the impact.